### DISCUSSION TITLE
blog/non-determinism-in-gpt-4/

### DISCUSSION THREAD

**Discussion by giscus** [152334H/152334H.github.io#8] [Announcements]
# blog/non-determinism-in-gpt-4/

It’s well-known at this point that GPT-4/GPT-3.5-turbo is non-deterministic, even at temperature=0.0. This is an odd behavior if you’re used to dense decoder-only models, where temp=0 should imply greedy sampling which should imply full determinism, because the logits for the next token should be a pure function of the input sequence & the model weights.

https://152334h.github.io/blog/non-determinism-in-gpt-4/

**Comment by joelxiangnanchen** [comment]
This blog post is really interesting! After reading it, I decided to browse OpenAI's public codebase to see if there were any tricks caused the similar case. Surprisingly, I found a trick in OpenAI's "Summarize from RLHF" codebase (https://github.com/openai/summarize-from-feedback/blob/8af822a428c93432aa80ffbe5b065a8f93895669/summarize_from_feedback/models/sample_fns.py#L20C47-L20C47). They added a small fraction to the temperature factor. We still don't know what other tricks they used. It's so weird!

**Reply by 152334H** [reply]
seems like that was added to lazily avoid a div-by-zero error && having a different code path for true greedy sampling.

Although i haven't checked, I would guess that a temperature of 1e-7 shouldn't result in a repeated sequence happening more than 1/30 times. But it does highlight the inconclusive role of the empirical experiment, as any weird implementation features (such as adding more temperature than 1e-7) could be causing the observations in the article.

**Reply by joelxiangnanchen** [reply]
Yes, it's interesting and we still don't reveal 'ClosedAI's magic'😂 And I'm curious about MoE based model's 'generation behavior feature'(E.g. some output diffs compared to the dense model) like this post demonstrated. But disappointed me that there are no other open-sourced llms based on MoE.😌

**Comment by joelxiangnanchen** [comment]
BTW, exactly, this is still a not-closed issue: (https://platform.openai.com/docs/guides/gpt/why-are-model-outputs-inconsistent) as they said: 'Setting temperature to 0 will make the outputs mostly deterministic, but a small amount of variability will remain.'

**Comment by hminooei** [comment]
In my understanding, the temp impacts the probabilities of tokens, but the sampling - which is controlled by top_p or top_k is still there hence the indeterminism even when temp=0.

**Comment by mdooley-9** [comment]
I guess I don't understand how the balancing out the load on the experts is actually causing a non-deterministic nature though. The only example I could fine of what it actually means I found here under the 'Reasonable Explaination' section of the article: https://medium.com/@toxa.ivchenko/gpt-4-non-deterministic-behavior-a-deep-dive-into-the-dark-mystery-373cbe683e4e

Assuming the example is right, the sequences of the words (in this case the A, B, C, and D's) would still be the same if the same input got sent through the model, unless there is a random component to the evening out of the load across the experts to my understanding. Where am I wrong?

**Reply by 152334H** [reply]
the "random component" comes in a potential expert capacity cutoff. in an expert parallel setup, you might want to assign each expert a "maximum proportion of total tokens processed per batch" (expert capacity), where some tokens will simply be no-op'd instead of processed when above capacity.

because exceeding that capacity is batch-dependent rather than only sequence-dependent, you have randomness across calls

**Reply by 152334H** [reply]
note: I don't think this is the correct answer for randomness in 2024. I think there are decent odds it was correct for its time, but at this point there are better engineering solutions than a hard token capacity cutoff.

**Comment by marhlder** [comment]
Could it be because OpenAI is failing to disable jitter noise in their MoE router during inference. Lol?

**Reply by 152334H** [reply]
no, and besides they won't be using TC in 2025

**Reply by marhlder** [reply]
"TC" ?

**Reply by 152334H** [reply]
I misthought "Token Capacity" instead of "Expert Capacity"
